src/ControlSystem/Averager.hpp

```diff
@@ -0,0 +1,77 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <deque>
+
+#include "DataStructures/DataVector.hpp"
+
+/// \ingroup ControlSystemGroup
+/// A weighted exponential averager of Q and its derivatives
+/// implementing Appendix A in Hemberger et al, https://arxiv.org/abs/1211.6079
+/// The averager is designed to support an arbitrary DerivOrder, however the
+/// private get_derivs method currently only supports DerivOrder = 2.
+/// If an additional DerivOrder is needed, finite differencing needs to be
+/// implemented to specifically handle that order (as it seems like overkill to
+/// generalize the differencing stencil at this time).
+template <size_t DerivOrder>
+class Averager {
+ public:
+  /// `avg_timescale_frac` determines the exponential averaging timescale
+  /// through \f$\tau_\mathrm{avg} = \f$`avg_timescale_frac`\f$\times \tau\f$,
+  /// where \f$\tau\f$ is the damping time.
+  /// `avg_timescale_frac` must be positive.
+  /// `average_0th_deriv_of_q` determines whether the call operator returns an
```

Comment:
This option comes from SpEC and is an input file option. It is not part of the call operator because it should not change once the decision whether to average the 0th derivative is made. From today's review they decided it is best not to make this a template parameter.

Reply:
As I've said several times now to other people. Saying "Was discussed in some closed forum" is not the right way to develop an open source code. The reason for the conclusion should at the very least be in the reply to my question, and even better would be as a C++ comment in the code.

Reply:
Let me try (I was at said meeting, our usual wednesday meeting, as were @kidder and @wthrowe): `average_0th_deriv_of_q` is currently intended to be an input file parameter, which would go in the constructor and be filled in by Options.  In SpEC, this parameter is set to false for most control systems and true for size and shape control since the latter are more noisy.  This parameter does not change during a simulation (at least it doesn't in SpEC), so indeed it could be a template parameter instead of an input file option.  I can see arguments for either, and I don't really care too much which one it is. Will was worried that someday, we might want to change this parameter during a run (e.g. some parts of a run, like junk radiation or merger, may be noisier than other parts, and might warrant different averaging), and if we eventually decide to do this, the required code changes would be easier if this parameter were a member variable than if it were a template parameter.  On that basis it seemed to us that a member variable was a marginally better choice. @wthrowe and @kidder feel free to add anything else you remember of the discussion.

Reply:
@nilsdeppe, thanks for the review. Apologies for the brief explanation above -- I had to leave suddenly in the middle of writing that (so I didn't elaborate too much). I also thought that our decision to keep it as an input file parameter was reason enough for not making it a template parameter (but I could have been clearer). In any case, I am on the same page as @markscheel (who did a much better job at explaining things) -- meaning I am fine doing it either way. Just let me know what you, @wthrowe and @kidder agree on and I'll do my best to implement it.